pygsti/models/model.py

```diff
@@ -1222,40 +1340,136 @@ def complete_circuit(self, circuit):
         ----------
         circuit : Circuit
             Circuit to act on.
-
+        
+        prep_lbl_to_prepend : Label, optional (default None)
+            Optional user specified prep label to prepend. If not
+            specified will use the default value as given by
+            :meth:_default_primitive_prep_layer_lbl. If the circuit
+            already has a prep label this argument will be ignored.
+
+        povm_lbl_to_append : Label, optional (default None)
+            Optional user specified prep label to prepend. If not
+            specified will use the default value as given by
+            :meth:_default_primitive_prep_layer_lbl. If the circuit
+            already has a prep label this argument will be ignored.
         Returns
         -------
         Circuit
             Possibly the same object as `circuit`, if no additions are needed.
         """
-        prep_lbl_to_prepend = None
-        povm_lbl_to_append = None
 
         if len(circuit) == 0 or not self._is_primitive_prep_layer_lbl(circuit[0]):
             prep_lbl_to_prepend = self._default_primitive_prep_layer_lbl()
             if prep_lbl_to_prepend is None:
-                #raise ValueError(f"Missing state prep in {circuit.str} and there's no default!")
-                raise ValueError("Missing state prep in %s and there's no default!" % circuit.str)
+                raise ValueError(f"Missing state prep in {circuit.str} and there's no default!")
 
         if len(circuit) == 0 or not self._is_primitive_povm_layer_lbl(circuit[-1]):
             sslbls = circuit.line_labels if circuit.line_labels != ("*",) else None
             povm_lbl_to_append = self._default_primitive_povm_layer_lbl(sslbls)
 
             if povm_lbl_to_append is None:
-                #raise ValueError(f"Missing POVM in {circuit.str} and there's no default!")
-                raise ValueError("Missing POVM in %s and there's no default!" % circuit.str)
-
-        if prep_lbl_to_prepend or povm_lbl_to_append:
-            #SLOW way:
-            #circuit = circuit.copy(editable=True)
-            #if prep_lbl_to_prepend: circuit.insert_layer_inplace(prep_lbl_to_prepend, 0)
-            #if povm_lbl_to_append: circuit.insert_layer_inplace(povm_lbl_to_append, len(circuit))
-            #circuit.done_editing()
-            if prep_lbl_to_prepend: circuit = (prep_lbl_to_prepend,) + circuit
-            if povm_lbl_to_append: circuit = circuit + (povm_lbl_to_append,)
+                raise ValueError(f"Missing POVM in {circuit.str} and there's no default!")
+                
+        if prep_lbl_to_prepend: 
+            circuit = (prep_lbl_to_prepend,) + circuit
+        if povm_lbl_to_append: 
+            circuit = circuit + (povm_lbl_to_append,)
 
         return circuit
 
+    def complete_circuits(self, circuits, prep_lbl_to_prepend=None, povm_lbl_to_append=None, return_split = False):
```

Comment:
We want to avoid calling the single case many times (the various model lookups for default POVMS and preps and whatnot genuinely takes a long time), but I think refactoring the single-use case to use the multi-circuit implementation should be fine.